### PR TITLE
bugfix: Rely respective TS type for IDL

### LINKED ISSUE #1: Error when running the frontend
This is an early attempt at making `create-solana-dapp` work with Anchor v0.30.0.

This repo is based on the latest `create-solana-dapp` and I made a few commits to make the `anchor` library work. 

`pnpm anchor build` and `pnpm anchor test` work.

However, the frontend does not seem to be happy.

image

There's a bounty for the person who fixes this.

### PR DESCRIPTION
**Pull request by sunguru98**
This pull request was created by https://app.gib.work/i/l2HzmWge/error-when-running-the-frontend in an attempt to solve a bounty. Payment for the bounty is immediately sent to the contributor after the merge.
 
 Closes #1

### CODE REVIEW

**Comment by sunguru98** [comment]
Posting the reason for the fix I've made ser @beeman. 

## TLDR

Classic JSON TS-type misunderstanding. lol

## Long explanation

So as you can see in the error message, it says the following
```txt
Type 'string' is not assignable to type 'IdlType'.ts(2345)
```

So, I went around and dug with the JSON IDL. Let me paste a small JSON instruction snippet from the IDL of why this error occurs. 

```json
{
  "name": "set",
  "discriminator": [
    198,
    51,
    53,
    241,
    116,
    29,
    126,
    194
  ],
  "accounts": [
    {
      "name": "app_1713693601",
      "writable": true
    }
  ],
  "args": [
    {
      "name": "value",
      "type": "u8"
    }
  ]
}
```
This IDL snippet is based on the `set` instruction. The issue lies in the `args[0].type`. Let's put a pin on this, as I'll come back to this in a bit, but just for now keep in mind that it's a normal string (because JSON can't be inferred as is, but can just be taken as the type of the value "u8"), and the issue lies in a field name called "type" inside "args", inside "instruction".

Let's now take a look at the type `IdlInstruction` in Anchor

```typescript
export type IdlInstruction = {
    name: string;
    docs?: string[];
    discriminator: IdlDiscriminator;
    accounts: IdlInstructionAccountItem[];
    args: IdlField[];
    returns?: IdlType;
};
```
We could see a field called `args`, which is the place the issue is arising. Let's now take a look at `IdlField` type.

```typescript
export type IdlField = {
    name: string;
    docs?: string[];
    type: IdlType;
};
```

Great, another field called type lol. Let's take a look at `IdlType`
```typescript
export type IdlType = "bool" | "u8" | "i8" | "u16" | "i16" | "u32" | "i32" | "f32" | "u64" | "i64" | "f64" | "u128" | "i128" | "u256" | "i256" | "bytes" | "string" | "pubkey" | IdlTypeOption | IdlTypeCOption | IdlTypeVec | IdlTypeArray | IdlTypeDefined | IdlTypeGeneric;
```

I'm sure by now, you could have guessed why that error is arising.
The `type` field doesn't have of TS type `string` anywhere inside `IdlType`. I'm stressing the TS type string again here, because "string" is not the same as string. "string" represents Rust's argument of `String`, but the string that I'm mentioning is of normal TS type `string`. Let me write it as a typescript snippet down to differentiate what I mean ser.

Expected type
```typescript
export type IdlType = string
```
The one present right now
```typescript
export type IdlType = "string"
```

Hope that makes sense ser.

So, when we directly infer from a JSON file, we don't take into the constraints of TS, but just take a raw value type (which would be of string)

The reason why the tests don't have this mess is because of the following line.
```typescript
const program = anchor.workspace.App1713693601 as Program<App1713693601>
```

If you notice carefully, we aren't inferring a JSON file here, but from a workspace (Anchor.toml) type. Since we are getting directly from a type, it automatically kind of gets from the TS file. 

This is why the following fix was made in the front end as well
```typescript
const program = new Program(App1713693601IDL as App1713693601, provider);
```

Now from 0.30, we don't have to infer program ID as an arg, as the IDL now has the address inferred in a more verbose manner under the metadata field (in JSON). The second thing is, we are now typecasting for TS to not get pissed off, by saying "Hey whatever I imported from JSON, typecast it to the right TS type, so that the IDLType is inferred right".


I might be wrong here, but afaik, this did the trick and was able to initialize accounts and set values :)

PS: The program ID is different due to the keypair not being present for the program ID (but I know you know this already ser haha)

Screenshot 2024-04-21 at 6 38 42 PM

**Comment by beeman** [comment]
Thanks @sunguru98! 🙌
